service/controller/clusterapi/v17/resources/clusterstatus/create.go

```diff
@@ -2,8 +2,242 @@ package clusterstatus
+	var tenantAPIAvailable bool
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding nodes of tenant cluster")
+
+		l, err := cc.Client.TenantCluster.K8s.CoreV1().Nodes().List(metav1.ListOptions{})
+		if tenantcluster.IsTimeout(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "timeout fetching certificates")
+		} else if tenant.IsAPINotAvailable(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "tenant API not available")
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			tenantAPIAvailable = true
+			nodes = l.Items
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d nodes from tenant cluster", len(nodes)))
+		}
+	}
+
+	var machineDeployments []cmav1alpha1.MachineDeployment
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding MachineDeployments for tenant cluster")
+
+		machineDeployments, err = r.getMachineDeployments(ctx, cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d MachineDeployments for tenant cluster", len(machineDeployments)))
+	}
+
+	updatedStatus = r.computeClusterConditions(ctx, cr, updatedStatus, tenantAPIAvailable, nodes, machineDeployments)
+
+	if !reflect.DeepEqual(currentStatus, updatedStatus) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updating cluster status")
+
+		cr = r.commonClusterStatusAccessor.SetCommonClusterStatus(cr, updatedStatus)
+		_, err = r.cmaClient.ClusterV1alpha1().Clusters(cr.Namespace).Update(&cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "updated cluster status")
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling reconciliation")
+		reconciliationcanceledcontext.SetCanceled(ctx)
+
+		return nil
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured cluster status is up to date")
+
 	return nil
 }
+
+func (r *Resource) ensureClusterHasID(ctx context.Context, cluster cmav1alpha1.Cluster, status v1alpha1.CommonClusterStatus) (v1alpha1.CommonClusterStatus, error) {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding out if cluster status has ID")
+
+	if status.ID != "" {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found out cluster has ID %s", status.ID))
+		return status, nil
+	}
+
+	clusterID := cluster.Labels[label.Cluster]
+	if clusterID != "" {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found out cluster has cluster label containing %s; reusing that", status.ID))
+		status.ID = clusterID
+		return status, nil
+	}
+
+	return status, microerror.Maskf(notFoundError, "cluster ID")
+}
+
+func (r *Resource) computeClusterConditions(ctx context.Context, cluster cmav1alpha1.Cluster, clusterStatus v1alpha1.CommonClusterStatus, tenantAPIAvailable bool, nodes []corev1.Node, machineDeployments []cmav1alpha1.MachineDeployment) v1alpha1.CommonClusterStatus {
```

Comment:
I would like to have a unit test for this function or re-write the conditions in a way that guarantee there is only one action taken.

The different block logic here make reasoning about the conditions hard I feel. I would prefer to gather in the first places all the elements needed to take a decision and have one if/else block in the end which show how the decision is taken instead of having those split up into different blocks.

Reply:
After getting this more or less ready, I was thinking the same. I'll see how I can change this so that it's a bit clearer. Initially I was about to add returns on conditions, but version management below is required occasionally and return would break that. I'll come back to you a bit later on this. :+1: